pkg/server/api/agent/v1/service_test.go

```diff
@@ -24,10 +30,95 @@ func (c *serviceTest) Cleanup() {
 	c.done()
 }
 
+func TestCreateJoinToken(t *testing.T) {
+	for _, tt := range []struct {
+		name          string
+		request       *agentpb.CreateJoinTokenRequest
+		expectLogs    []spiretest.LogEntry
+		expectResults *types.JoinToken
+		err           string
+		code          codes.Code
+		dsError       error
+	}{
+		{
+			name: "Success Basic Create Join Token",
+			request: &agentpb.CreateJoinTokenRequest{
+				Ttl: 1000,
+			},
+		},
+		{
+			name: "Success Custom Value Join Token",
+			request: &agentpb.CreateJoinTokenRequest{
+				Ttl:   1000,
+				Token: "token goes here",
+			},
+		},
+		{
+			name: "Fail Negative Ttl",
+			request: &agentpb.CreateJoinTokenRequest{
+				Ttl: -1000,
+			},
+			err:  "ttl is required, you must provide one",
+			code: codes.InvalidArgument,
+		},
+		{
+			name: "Fail Datastore Error",
+			err:  "Failed to create token: datatore broken",
+			request: &agentpb.CreateJoinTokenRequest{
+				Ttl: 1000,
+			},
+			dsError: errors.New("datatore broken"),
+			code:    codes.Internal,
+		},
+	} {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			test := setupServiceTest(t)
+			test.ds.SetNextError(tt.dsError)
+
+			result, err := test.client.CreateJoinToken(context.Background(), tt.request)
+			if tt.err != "" {
+				require.Error(t, err, tt.err)
+				spiretest.RequireGRPCStatusContains(t, err, tt.code, tt.err)
+			} else {
+				require.NoError(t, err)
+				require.NotNil(t, result)
+				require.NotEmpty(t, result.Value)
+				require.NotEmpty(t, result.Value)
+			}
+		})
+	}
+}
+
+func TestCreateJoinTokenWithAgentId(t *testing.T) {
+	test := setupServiceTest(t)
+
+	_, err := test.client.CreateJoinToken(context.Background(), &agentpb.CreateJoinTokenRequest{
+		Ttl:     1000,
+		AgentId: &types.SPIFFEID{TrustDomain: "badtd.org", Path: "invalid"},
+	})
+	require.Error(t, err)
+	spiretest.RequireGRPCStatusContains(t, err, codes.InvalidArgument, "requested agent SPIFFE ID does not match server trust domain")
+
+	token, err := test.client.CreateJoinToken(context.Background(), &agentpb.CreateJoinTokenRequest{
+		Ttl:     1000,
+		AgentId: &types.SPIFFEID{TrustDomain: "example.org", Path: "valid"},
+	})
+	require.NoError(t, err)
+	spiretest.RequireGRPCStatusContains(t, err, codes.OK, "")
+
+	listEntries, err := test.ds.ListRegistrationEntries(context.Background(), &datastore.ListRegistrationEntriesRequest{})
+	require.NoError(t, err)
+	require.Equal(t, "spiffe://example.org/spire/agent/join_token/"+token.Value, listEntries.Entries[0].ParentId)
+	require.Equal(t, "spiffe://example.org/valid", listEntries.Entries[0].Selectors[0].Value)
+}
+
 func setupServiceTest(t *testing.T) *serviceTest { //nolint: unused,deadcode
 	ds := fakedatastore.New(t)
+	serverTrustDomain := spiffeid.RequireTrustDomainFromString("example.org")
```

Comment:
nit: inline this when creating the `Config` struct since it isn't used elsewhere?